File: haskell/src/Wildling/Generator.hs
module Wildling.Generator
  ( Generator(..)
  , createGenerator
  , generatorCount
  , generatorGet
  ) where

import Wildling.ParsePattern (Dictionaries, parsePattern)
import Wildling.Token (Token, tokenCount, tokenGet)

data Generator = Generator
  { generatorSource :: String
  , generatorTokens :: [Token]
  , generatorCount_ :: Int
  }

createGenerator :: String -> Dictionaries -> Generator
createGenerator inputPattern dictionaries =
  let tokens = parsePattern inputPattern dictionaries
      count = product (map tokenCount tokens)
   in Generator
        { generatorSource = inputPattern
        , generatorTokens = tokens
        , generatorCount_ = count
        }

generatorCount :: Generator -> Int
generatorCount = generatorCount_

generatorGet :: Generator -> Int -> String
generatorGet gen index
  | index > generatorCount_ gen - 1 || index < 0 = ""
  | otherwise = concat (go index (generatorTokens gen))
  where
    go _ [] = []
    go idx (t : ts) =
      let c = tokenCount t
       in tokenGet t (idx `mod` c) : go (idx `div` c) ts

File: go/wildling/generator.go
package wildling

// Generator expands a single pattern.
type Generator struct {
	Source string
	tokens []*Token
	count  int
}

// NewGenerator creates a generator for one pattern.
func NewGenerator(inputPattern string, dictionaries Dictionaries) *Generator {
	tokens := ParsePattern(inputPattern, dictionaries)
	count := 1
	for _, token := range tokens {
		count *= token.Count()
	}
	return &Generator{
		Source: inputPattern,
		tokens: tokens,
		count:  count,
	}
}

func (g *Generator) Count() int {
	return g.count
}

func (g *Generator) Tokens() []*Token {
	return g.tokens
}

func (g *Generator) Get(index int) string {
	if index > g.count-1 || index < 0 {
		return ""
	}
	parts := make([]string, len(g.tokens))
	indexWithOffset := index
	for i, token := range g.tokens {
		parts[i] = token.Get(indexWithOffset % token.Count())
		indexWithOffset /= token.Count()
	}
	out := ""
	for _, p := range parts {
		out += p
	}
	return out
}
